src/ng/compile.js

```diff
@@ -3500,7 +3500,9 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
       });
 
       function recordChanges(key, currentValue, previousValue) {
-        if (isFunction(destination.$onChanges) && currentValue !== previousValue) {
+        if (isFunction(destination.$onChanges) && currentValue !== previousValue &&
+            // eslint-disable-next-line no-self-compare
+            (currentValue === currentValue || previousValue === previousValue)) {
```

Comment:
I'll add the extra test. It is good excuse to fix [an indentation issue][1] :stuck_out_tongue: 

[1]: https://github.com/angular/angular.js/commit/7d7efbf545c8c07713eb45301660dcfca4121445#diff-348c2f3781ed66a24894c2046a52c628R4400